.github/workflows/build_test_and_deploy.yml

```
@@ -0,0 +1,91 @@
+# Copyright (c) 2022 ETH Zurich
+#
+# SPDX-License-Identifier: BSL-1.0
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+name: Build, test and deploy docker images
+
+# The builds are triggered with pushes on the 'main' branch or with a pull
+# request against the 'main' branch
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  # Build base docker image
+  build_base_image:
+    env:
+      version: 1
+      image_name: pika-ci-base
+    # Cannot use the env variables in the name
+    name: github/pika-ci-base
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USER }}
+          password: ${{ secrets.DOCKER_PASS }}
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      # Pull request target is used to match all events on the pull request
+      - name: Build 
+        if: ${{ github.event != 'push' }}
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: Dockerfile.base
+          push: false
+          tags: ${{ secrets.DOCKER_USER }}/${{ env.image_name }}:latest,${{ secrets.DOCKER_USER }}/${{ env.image_name }}:${{ env.version }}
+      # The deploy step is only triggered on push to main branch
+      - name: Build and deploy
+        if: ${{ github.event == 'push' }}
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: Dockerfile.base
+          push: true
+          tags: ${{ secrets.DOCKER_USER }}/${{ env.image_name }}:latest,${{ secrets.DOCKER_USER }}/${{ env.image_name }}:${{ env.version }}
+
+  # Build HIP docker image
+  build_hip_image:
+    env:
+      version: 1
+      image_name: pika-ci-hip
+    # Cannot use the env variables in the name
+    name: github/pika-ci-hip
+    runs-on: ubuntu-latest
+    needs: build_base_image
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKER_USER }}
+          password: ${{ secrets.DOCKER_PASS }}
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      # Pull request target is used to match all events on the pull request
+      - name: Build 
+        if: ${{ github.event != 'push' }}
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: Dockerfile.base
+          push: false
+          tags: ${{ secrets.DOCKER_USER }}/${{ env.image_name }}:latest,${{ secrets.DOCKER_USER }}/${{ env.image_name }}:${{ env.version }}
```

Maybe?
```suggestion
          tags: ${{ secrets.DOCKER_USER }}/${{ env.image_name }}:latest,${{ secrets.DOCKER_USER }}/${{ env.image_name }}:${{ env.version }}
          build-arg: BASE_IMAGE=${{ secrets.DOCKER_USER }}/${{ env.image_name }}:${{ env.version }}
```
It might be able to use the non-pushed image with the given versioned name. Alternatively, you could also have a `tmp` tag or something like that which is only used in this non-deploy build.